packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js

```diff
@@ -0,0 +1,454 @@
+  makeRatioFromAmounts,
+  floorDivideBy,
+} from '@agoric/zoe/src/contractSupport/index.js';
+
+const A = 85;
+const BASIS_POINTS = 10000n;
+
+const within10 = (a, b) => {
+  if (a > b) {
+    return a - b <= 10;
+  }
+  return b - a <= 10;
+};
+
+/**
+ * Recompute the pool values after a swap is performed
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ *                                passed from the contract.
+ * @param {number} tokenIndexFrom - index of amount of inputReserve
+ *                                  in the reserves array.
+ * @param {number} tokenIndexTo - index of amount of outputReserve
+ *                                in the reserves array.
+ * @param {bigint} inputTokenValue - the value of swap in token passed in.
+ * @param {bigint} outputTokenValue - the value of swap out token returned
+ *                                    in exchange
+ * @returns {bigint[]} output - recomputed pool values
+ */
+const updatePoolValues = (
+  poolValues,
+  tokenIndexFrom,
+  inputTokenValue,
+  tokenIndexTo,
+  outputTokenValue,
+) => {
+  poolValues[tokenIndexFrom] -= inputTokenValue;
+  poolValues[tokenIndexTo] += outputTokenValue;
+  return poolValues;
+};
+
+/**
+ * Computes the Stable Swap invariant (D).
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ * passed from the contract.
+ * @returns {bigint} d - the current price, in value form
+ *
+ */
+const getD = poolValues => {
+  let N_COINS = poolValues.length;
+  let sum_x = 0n;
+
+  // sum_x - Sum of all poolValues.
+  for (let i = 0; i < N_COINS; i++) {
+    sum_x = sum_x + poolValues[i];
+  }
+  if (sum_x === 0n) {
+    return 0n;
+  }
+  let d_prev = 0n;
+  let d = sum_x;
+  let nA = A * N_COINS;
+
+  for (let i = 0; i < 1000; i++) {
+    let dp = d;
+    // prod - product of all poolvalues
+    // dp = D^(n+1)/n^n(prod)
+    for (let j = 0; j < N_COINS; j++) {
+      dp = (dp * d) / (poolValues[j] * Nat(N_COINS));
+    }
+
+    d_prev = d;
+
+    // numerator = An(sum) + ((D^(n+1)/n^n(prod))*nD)
+    // denominator = ((An-1)*D)+(n+1)(D^(n+1)/n^n(prod))
+    // d = numerator/denominator
+    d =
+      ((Nat(nA) * sum_x + dp * Nat(N_COINS)) * d) /
+      ((Nat(nA) - 1n) * Nat(d) + Nat(N_COINS + 1) * dp);
+    if (within10(d, d_prev)) {
+      return d;
+    }
+  }
+  return d;
+};
+
+/**
+ * Compute the swap amount `y` in proportion to `x`.
+ *
+ * @param {bigint} x - index of amount of inputReserve
+ * in the reserves array.
+ * @param {number} tokenIndexFrom - index of amount of inputReserve
+ * in the reserves array.
+ * @param {number} tokenIndexTo - index of amount of outputReserve
+ * in the reserves array.
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ * passed from the contract.
+ * @returns {bigint} y - the amount of swap out asset to be returned
+ * in exchange for amount x of swap in asset.
+ */
+
+const getY = (x, tokenIndexFrom, tokenIndexTo, poolValues) => {
```

Comment:
I don't understand where these equations come from in the paper or anywhere else. Can you explain?
```
  // c=(D^(n+1))/(n^n)*prod`
  // c = ([(D^(n+1))/(n^n)*prod`]*d)/(A(n^n))
  // b = s/(D*An)
  // numerator = ((y^2)+([(D^(n+1))/(n^n)*prod`]*d)/(A(n^n)))
  // denominator = 2y+ (s/(D*An)) - D
 ```

Reply:
- This is quite simplified form of Yi+1 which is calculated in the similar manner as Di+1.
- c and b are some sub expression
- I am working on it will shift the formula to non simplified form just as I did for D.
- The formula will be equivalent in terms of value but different expression.
- Soon will provide detailed calculation of this formula as well.

Reply:
### Formula Calculations for Y:
<p align="left">
<img src="https://user-images.githubusercontent.com/55734494/155235661-b49349e6-2a69-4929-bee9-707b6cf7f1c6.jpeg"  width="850" height="850"/>
<p>

- Previously a very simplified form of yi+1 was being used.
- Have shifted the formula to the formula provided above for better understanding and ease.